File: app/src/main/java/md/fusionworks/lifehack/util/rx/ObservableTransformation.java
package md.fusionworks.lifehack.util.rx;

import java.io.IOException;

import md.fusionworks.lifehack.data.api.ResponseCode;
import md.fusionworks.lifehack.data.api.exception.NotFoundException;
import md.fusionworks.lifehack.data.api.exception.UnknownException;
import md.fusionworks.lifehack.util.Constant;
import retrofit2.Response;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class ObservableTransformation {

    private static Observable.Transformer ioToMainThreadSchedulerTransformer;
    private static Observable.Transformer apiRequestConfigurationTransformer;

    static {
        ioToMainThreadSchedulerTransformer = createIOToMainThreadScheduler();
        apiRequestConfigurationTransformer = createApiRequestConfigurationTransformer();
    }

    @SuppressWarnings("unchecked")
    private static <T> Observable.Transformer<T, T> createIOToMainThreadScheduler() {
        return tObservable -> tObservable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @SuppressWarnings("unchecked")
    private static <T> Observable.Transformer<Response<T>, T> createApiRequestConfigurationTransformer() {
        return responseObservable -> responseObservable
                .flatMap(tResponse -> Observable.create((Observable.OnSubscribe<Response<T>>) subscriber -> {
                    if (tResponse.isSuccess()) {
                        subscriber.onNext(tResponse);
                        subscriber.onCompleted();
                    } else {
                        if (tResponse.code() == ResponseCode.NOT_FOUND)
                            subscriber.onError(new NotFoundException());
                        else
                            subscriber.onError(new UnknownException());
                    }
                }))
                .retryWhen(new RetryWithDelayIf(Constant.MAX_RETRIES, Constant.RETRY_DELAY_MILLIS, throwable -> throwable instanceof IOException))
                .map(t -> t.body());
    }

    @SuppressWarnings("unchecked")
    public static <T> Observable.Transformer<T, T> applyIOToMainThreadSchedulers() {
        return ioToMainThreadSchedulerTransformer;
    }

    @SuppressWarnings("unchecked")
    public static <T> Observable.Transformer<Response<T>, T> applyApiRequestConfiguration() {
        return apiRequestConfigurationTransformer;
    }
}

File: app/src/main/java/md/fusionworks/lifehack/util/rx/ObservableTransformation.kt
package md.fusionworks.lifehack.util.rx

import md.fusionworks.lifehack.data.api.NotFoundException
import md.fusionworks.lifehack.data.api.ResponseCode
import md.fusionworks.lifehack.data.api.UnknownException
import md.fusionworks.lifehack.util.Constant
import retrofit2.Response
import rx.Observable
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.io.IOException

object ObservableTransformation {

  @SuppressWarnings("unchecked")
  fun <T> applyIOToMainThreadSchedulers(): Observable.Transformer<T, T> = Observable.Transformer<T, T> { tObservable ->
    tObservable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
  }

  @SuppressWarnings("unchecked")
  fun  <T> applyApiRequestConfiguration(): Observable.Transformer<Response<T>, T> {
    return Observable.Transformer<Response<T>, T> {
      it.flatMap { response ->
        Observable.create<Subscriber<Response<T>>> { it ->
          Observable.OnSubscribe<Response<T>> {
            if (response.isSuccessful) {
              it.onNext(response)
              it.onCompleted()
            } else {
              if (response.code() == ResponseCode.NOT_FOUND) {
                it.onError(NotFoundException())
              } else {
                it.onError(UnknownException())
              }
            }
          }
        }
      }
      it.retryWhen(
          RetryWithDelayIf(Constant.MAX_RETRIES, Constant.RETRY_DELAY_MILLIS) { it is IOException })
          .map { it.body() }
    }
  }
}